json/test_rapid_sax.cpp

```diff
@@ -21,83 +21,106 @@ struct coordinate_t {
   double y;
   double z;
 
-  auto operator<=>(const coordinate_t&) const = default;
+  auto operator<=>(const coordinate_t &) const = default;
 
-  friend ostream& operator<< (ostream &out, const coordinate_t &point) {
-    out << "coordinate_t {x: " << point.x
-        << ", y: " << point.y
+  friend ostream &operator<<(ostream &out, const coordinate_t &point) {
+    out << "coordinate_t {x: " << point.x << ", y: " << point.y
         << ", z: " << point.z << "}";
     return out;
   }
 };
 
-using TCallback = function<void(const coordinate_t&)>;
+using TCallback = function<void(const coordinate_t &)>;
 
 class CoordinateHandler : public BaseReaderHandler<UTF8<>, CoordinateHandler> {
 public:
-  CoordinateHandler(const TCallback& callback):
-    state_(kStart), x_(), y_(), z_(), callback_(callback) {}
+  CoordinateHandler(const TCallback &callback)
+      : state_(kStart), x_(), y_(), z_(), callback_(callback) {}
 
   bool Double(double d) {
     switch (state_) {
-      case kX: x_ += d; state_ = kCoordinatesElement; break;
-      case kY: y_ += d; state_ = kCoordinatesElement; break;
-      case kZ: z_ += d; state_ = kCoordinatesElement; break;
-      default: break;
+    case kX:
+      x_ += d;
+      state_ = kCoordinatesElement;
+      break;
+    case kY:
+      y_ += d;
+      state_ = kCoordinatesElement;
+      break;
+    case kZ:
+      z_ += d;
+      state_ = kCoordinatesElement;
+      break;
+    default:
+      break;
     }
     return true;
   }
 
   bool StartObject() {
     switch (state_) {
-      case kStart: state_ = kRoot; break;
-      case kCoordinatesArray: state_ = kCoordinatesElement; break;
-      default: break;
+    case kStart:
+      state_ = kRoot;
+      break;
+    case kCoordinatesArray:
+      state_ = kCoordinatesElement;
+      break;
+    default:
+      break;
     }
     return true;
   }
 
-  bool Key(const Ch* str, SizeType len, bool) {
+  bool Key(const Ch *str, SizeType len, bool) {
     switch (state_) {
-      case kRoot:
-        if (len == sizeof("coordinates") - 1 && memcmp(str, "coordinates", sizeof("coordinates") - 1) == 0)
-          state_ = kCoordinates;
-        break;
+    case kRoot:
+      if (len == sizeof("coordinates") - 1 &&
+          memcmp(str, "coordinates", sizeof("coordinates") - 1) == 0)
+        state_ = kCoordinates;
+      break;
 
-      case kCoordinatesElement:
-        if (len == 1 && str[0] >= 'x' && str[0] <= 'z')
-          state_ = static_cast<State>(kX + (str[0] - 'x'));
-        break;
+    case kCoordinatesElement:
+      if (len == 1 && str[0] >= 'x' && str[0] <= 'z')
```

Comment:
That's a deliberate hack I didn't notice before. With the hash helpers we could make it properly:

```cpp
switch (hash) {
  case "x"_hash: state_ = kX; break;
  case "y"_hash: state_ = kY; break;
  case "z"_hash: state_ = kZ; break;
}
```